### PR TITLE
Remove IDisposable from MessageSession and PipelineContextSession

### LINKED ISSUE #8: UniformSession disposed when container tracks disposables
Both the `MessageSession` and the `PipelineContextSession` implement `IDisposable` internally. Some containers like Autofac even detect this on internal types and track those disposables. When the child container is out of scope all the disposables will get disposed. 

## Steps to reproduce

1. Create AspNET Core WebApi
1. Use UniformSession
1. Use Autofac
1. Add following code

```
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddMvc();
            services.AddScoped<ReusedComponent>();
        }

    public class ReusedComponent
    {
        private readonly IUniformSession _session;

        public ReusedComponent(IUniformSession session)
        {
            _session = session;
        }

        public Task Do()
        {
            return _session.Publish(new DebugEvent());
        }
    }

    public class DebugEvent : IEvent { }

    [Route("api/debug")]
    public class DebugController : ControllerBase
    {
        private readonly ReusedComponent _component;

        public DebugController(ReusedComponent component)
        {
            _component = component;
        }

        public async Task<IActionResult> Post()
        {
            await _component.Do();

            return Ok();
        }
    }
```

Second call on the controller will lead to an `InvalidOperationException` indicating the session was disposed.

## Workaround

With autofac switch to

```
    public class ReusedComponent
    {
        private readonly IUniformSession _session;

        public ReusedComponent(Owned<IUniformSession> session)
        {
            _session = session.Value;
        }

        public Task Do()
        {
            return _session.Publish(new DebugEvent());
        }
    }
```

### PR DESCRIPTION
**Pull request by danielmarbach**
Closes #8 

Smoke tested with repro from @jenspettersson

smoke-test

### CODE REVIEW

File: src/AcceptanceTests/When_sending_from_cached_message_session_after_shutdown.When_sending_from_cached_message_session_after_shutdown.cs [hunk 1]
```diff
@@ -20,7 +20,7 @@ public class When_sending_from_cached_message_session_after_shutdown : NServiceB
 
             Assert.NotNull(ctx.StartupUniformSession);
             var exception = Assert.ThrowsAsync<InvalidOperationException>(() => ctx.StartupUniformSession.SendLocal(new MyMessage()));
-            StringAssert.Contains("This session has been disposed and can no longer send messages.", exception.Message);
+            StringAssert.Contains("This session has been marked as no longer usable and can no longer send messages.", exception.Message);
```
Comment (timbussmann) [inline src/AcceptanceTests/When_sending_from_cached_message_session_after_shutdown.When_sending_from_cached_message_session_after_shutdown.cs]: I prefer the previous sentence. I think it's valid to still describe the session as disposed, we're just using a different approach?

File: src/UniformSession/MessageSession.cs [hunk 2]
```diff
@@ -13,46 +13,46 @@ public MessageSession(IMessageSession messageSession, CurrentSessionHolder sessi
...
     {
-        ThrowIfDisposed();
+        ThrowIfUnusable();
         ThrowIfAbusedInsidePipeline();
 
         return messageSession.Publish(messageConstructor, publishOptions);
     }
 
-    public void Dispose()
+    public void MarkAsUnusable()
```
Comment (timbussmann) [inline src/UniformSession/MessageSession.cs]: maybe just `Close`?

File: src/UniformSession/MessageSession.cs [hunk 3]
```diff
@@ -66,8 +66,8 @@ void ThrowIfAbusedInsidePipeline()
 
     readonly IMessageSession messageSession;
 
-    bool isDisposed;
-    static readonly string AccessDisposedSessionExceptionMessage = $"This session has been disposed and can no longer send messages. Ensure to not cache instances {nameof(IUniformSession)}.";
-    static readonly string SessionUsedInsidePipelineExceptionMessage = $"This session is being used inside the message handling pipeline which can lead to message loss. Ensure to not cache instances {nameof(IUniformSession)}.";
+    bool closed;
     CurrentSessionHolder sessionHolder;
+    static readonly string AccessClosedSessionExceptionMessage = $"The endpoint owning this session instance has been stopped. It is no longer possible to execute message operations on this instance. Ensure to not cache instances of {nameof(IUniformSession)}.";
```
Comment (bording) [inline src/UniformSession/MessageSession.cs]: How about:

> The endpoint owning this {nameof(IUniformSession)} instance has been stopped, so it is no longer possible to execute message operations. 


"Ensure to not cache instances of {nameof(IUniformSession)}."
This part of the sentence is a bit awkwardly phrased, and it's not really clear what it's telling the users not to do. Is there a more specific action we can tell them to avoid?

File: src/UniformSession/MessageSession.cs [hunk 4]
```diff
@@ -66,8 +66,8 @@ void ThrowIfAbusedInsidePipeline()
...
+    static readonly string SessionUsedInsidePipelineExceptionMessage = $"This session is being used inside the message handling pipeline this can lead to message duplication. Ensure to not cache instances of {nameof(IUniformSession)}.";
```
Comment (bording) [inline src/UniformSession/MessageSession.cs]: How about:
> This {nameof(IUniformSession)} session is being used inside the message handling pipeline, which can lead to message duplication.

And same comment about the "Ensure" sentence. What leads to this exception being thrown? Can we give them a more specific action to take to fix the problem?

File: src/UniformSession/MessageSession.cs (hunk 3 above)
Comment (danielmarbach) [reply src/UniformSession/MessageSession.cs]: Maybe that addition is not even needed. Or should we rephrase it to

> Ensure to not use the message session after the endpoint was stopped.

File: src/UniformSession/MessageSession.cs (hunk 4 above)
Comment (danielmarbach) [reply src/UniformSession/MessageSession.cs]: How about

> Ensure  to not use the message session outside the message handling pipeline

File: src/UniformSession/MessageSession.cs (hunk 3 above)
Comment (bording) [reply src/UniformSession/MessageSession.cs]: I'd say that's redundant and can just be dropped instead.

File: src/UniformSession/MessageSession.cs (hunk 4 above)
Comment (bording) [reply src/UniformSession/MessageSession.cs]: That feels a bit redundant too. Maybe the second sentence isn't needed?
